### PR TITLE
[WIP] Implement a ParameterHandler::add_parameter() function.

### PR DESCRIPTION
**Pull request by bangerth**

### CODE REVIEW

**Comment by bangerth** [comment]
@luca-heltai -- here is a very rough draft, with no documentation and many missing pieces. The only thing that is important are the two last commits, the rest will disappear after merging #4312 and another intermediate PR. 

This currently works for the basic data types and `std::vector<std::string>`. It doesn't work right now with `std::vector<int>` or similar -- I'll have to think about this some more, and I realize that that is exactly where your solution in #4299 uses much of the complexity that I'm trying to avoid here.

File: include/deal.II/base/parameter_handler.h [hunk 1]
```diff
@@ -151,6 +154,8 @@ namespace Patterns
   class Integer : public PatternBase
   {
   public:
+    typedef int value_type;
```
Comment (luca-heltai) [inline include/deal.II/base/parameter_handler.h]: Do you want to create a new `Patterns::UnsignedInteger` for unsigned ints?

File: include/deal.II/base/parameter_handler.h [hunk 2]
```diff
@@ -225,6 +256,37 @@ namespace Patterns
...
+    };
+  };
+
+  template <>
+  class DefaultPattern<unsigned int>
+  {
+  public:
+    static Integer create()
+    {
+      return Integer();
```
Comment (luca-heltai) [inline include/deal.II/base/parameter_handler.h]: `return Integer(0);` should suffice...

**Comment by luca-heltai** [comment]
I don't see how this can be extended to use the same Pattern for `Point<3>` and `std::vector<double>`. Do you intend to implement a new pattern for every `std::vector<T>` type, and for every type we want to support in parameter handler? This requires many more lines of codes compared to my original proposal. 

We could merge the ideas here. Remove the `typedef` in the `Pattern` class, let the `to_string` and `to_value` be handled by the derived classes (if you really dislike it being in the base class), and let the `to_value` take two arguments as in my version, so that you can overload, for every pattern, whatever the pattern is compatible with.

With this, it would be possible to do 
~~~
template<typename T>
std:: string List::to_string(const std::vector<T> &value);

template<typename T>
std:: string List::to_string(const std::list<T> &value);

template<typename T>
std:: string List::to_string(const Vector<T> &value);

template<int dim>
std:: string List::to_string(const Point<dim> &value);

...

template<typename T>
void List::to_value(const std::string &s, std::vector<T> &value);

....
~~~

I understand you don't like the idea of simply having a function:

~~~
template<class T>
std::unique_ptr<PatternBase> Patterns::default_pattern();
~~~

but why? 

Is it really necessary to make a new templated class, with a public member function `create`?

File: include/deal.II/base/parameter_handler.h [hunk 3]
```diff
@@ -432,6 +557,66 @@ namespace Patterns
...
+     * The given value will be matched against the current pattern to make
+     * sure that it satisfies the constraints imposed (such as minimal and
+     * maximal values).
+     */
+    value_type  to_value (const std::string &s) const
+    {
+      std::string tmp = s;
+      std::vector<std::string> split_list;
+
+      // first split the input list
```
Comment (luca-heltai) [inline include/deal.II/base/parameter_handler.h]: why not using `Utilities::split_string_list`?

**Comment by luca-heltai** [comment]
On second thought, your implementation of `DefaultPattern<T>::create()` saves many lines of codes compared to mine.

**Comment by bangerth** [comment]
Made obsolete by #4632.

**Pull request closed without merging**